### PR TITLE
Implement custom signature verification level

### PR DESCRIPTION
**Pull request by rgnote**
Spec : https://github.com/notaryproject/notaryproject/pull/164

Signed-off-by: rgnote <5878554+rgnote@users.noreply.github.com>

### CODE REVIEW

**Comment by iamsamirzon** [comment]
@rgnote   - This is related to to the Trust Store and Trust Policy implementation. I am going to link it to the existing roadmap item https://github.com/notaryproject/roadmap/issues/45

File: verification/types.go [hunk 1]
```diff
@@ -120,13 +127,84 @@ var (
...
+	}
+	for key, value := range overrideMap {
+		var verificationType VerificationType
+		for _, t := range VerificationTypes {
+			if strings.EqualFold(string(t), key) {
+				verificationType = t
+			}
+		}
+		if verificationType == "" {
+			return nil, fmt.Errorf("verification type %q in custom signature verification is not supported", verificationType)
```
Comment (binbin-li) [inline verification/types.go:188]: I feel `key` should be passed in instead of the empty `verificationType`

File: verification/types.go [hunk 2]
```diff
@@ -120,13 +127,84 @@ var (
...
+		}
+
+		var verificationAction VerificationAction
+		for _, action := range VerificationActions {
+			if strings.EqualFold(string(action), value.(string)) {
+				verificationAction = action
+			}
+		}
+		if verificationAction == "" {
+			return nil, fmt.Errorf("verification action %q in custom signature verification is not supported", verificationAction)
```
Comment (binbin-li) [inline verification/types.go:198]: Same here

File: verification/types.go (hunk 1 above)
Comment (rgnote) [reply verification/types.go]: Good catch. Addressed in the latest revision

File: verification/types.go [hunk 3]
```diff
@@ -120,13 +121,67 @@ var (
...
+	for k, v := range baseLevel.VerificationMap {
+		customVerificationLevel.VerificationMap[k] = v
+	}
+
+	// override the verification actions with the user configured settings
+	for key, value := range signatureVerification.Override {
+		var verificationType VerificationType
+		for _, t := range VerificationTypes {
+			if strings.EqualFold(string(t), key) {
+				verificationType = t
```
Comment (priteshbandi) [inline verification/types.go:161]: should we break the loop when we have found a match?

File: verification/types.go [hunk 4]
```diff
@@ -120,13 +121,67 @@ var (
...
+			}
+		}
+		if verificationType == "" {
+			return nil, fmt.Errorf("verification type %q in custom signature verification is not supported", key)
+		}
+
+		var verificationAction VerificationAction
+		for _, action := range VerificationActions {
+			if strings.EqualFold(string(action), value) {
```
Comment (priteshbandi) [inline verification/types.go:170]: should we break the loop when we have found a match?

**Review by priteshbandi — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Comment by dtzar** [comment]
Per the [PR review criteria](https://github.com/notaryproject/notation/discussions/263) @notaryproject/notation-sub-group-maintainers - this is past a week old and has two approvals from official code-reviewers, so it is eligible for merge now.

@rgnote Just needs merge conflict resolved.

**Comment by priteshbandi** [comment]
LGTM.

**Review by priteshbandi — APPROVED** [review]

**Review by gokarnm — APPROVED** [review]
LGTM!

**Pull request merged**